### ISSUE TITLE
BlenderBot2 CUDA Out-Of-Memory(OOM) with 11Gb gpu 

### ISSUE THREAD

**Issue by WindChimeRan** [facebookresearch/ParlAI#4041]
Hi. I am running blenderbot2_400M model on 1080ti-11Gb

I have tested blenderbot2_400M without internet, which is ok. memory use: 6707 / 11175 MB

Then I tried the complete blenderbot2_400M with an internet search engine, when the searching is triggered, my machine reported OOM error.

It seems like the blenderbot2_400M model has two BART decoders (query_generator, memory_decoder) and my machine cannot afford two. 

Is there any workaround to reduce the memory consumption?

Thanks

**Comment by mojtaba-komeili** [comment]
Are you training a model, or running them in interactive mode. If you are training (or evaluating) a model, one way to avoid this OOM is to decrease your batch size. Another thing you can do is truncating your input text (`--text-truncate`) or your generated text (`--label-truncate`).

**Comment by WindChimeRan** [comment]
I am deploying the model via chat_service. It is mostly a terminal chat service. https://parl.ai/docs/tutorial_chat_service.html#terminal

**Comment by WindChimeRan** [comment]
```yml
tasks:
  default:
    onboard_world: MessengerBotChatOnboardWorld
    task_world: MessengerBotChatTaskWorld
    timeout: 18000
    agents_required: 1
task_name: chatbot
world_module: parlai.chat_service.tasks.chatbot.worlds
overworld: MessengerOverworld
max_workers: 30
opt:
  debug: True
  models:
    blenderbot2_400M:
      model:  projects.blenderbot2.agents.blenderbot2:BlenderBot2FidAgent
      model_file: zoo:blenderbot2/blenderbot2_400M/model
      interactive_mode: True
      no_cuda: True
      override:
        search_server: 0.0.0.0:8080
additional_args:
  page_id: 1 # Configure Your Own Page
```
here is my yml file. I also changed the MODEL_KEY from  'blender_90M' to 'blenderbot2_400M' in ParlAI/parlai/chat_service/tasks/chatbot/worlds.py to make MessengerBotChatTaskWorld works.

although `no_cuda: True`, from my observation on nvidia-smi, the model was still using GPU ...


Thanks!

**Comment by stephenroller** [comment]
Hm I really wouldn't expect you to have memory issues at deploy time -- people have deployed much larger models on similar GPUs.

Do you happen to see anything in the logs about the optimizer being loaded?

**Comment by stephenroller** [comment]
Also, you are running with no_cuda: True, meaning I don't think you're using the GPU at all

**Comment by WindChimeRan** [comment]
```bash
python search_server.py serve --host 0.0.0.0:8080
parlai interactive -mf zoo:blenderbot2/blenderbot2_400M/model --search_server 0.0.0.0:8080 --share-search-and-memory-query-encoder True --query-generator-truncate 256 --truncate 256 --text-truncate 256 --label-truncate 256 --checkpoint-activations true
```
I tried different truncate with interactive mode.
It works! memory consumption: 8436 / 11175 MB

**Comment by WindChimeRan** [comment]
How can I write the config file for chat_service?

Here is my current yml file:
```yaml
tasks:
  default:
    onboard_world: MessengerBotChatOnboardWorld
    task_world: MessengerBotChatTaskWorld
    timeout: 18000
    agents_required: 1
task_name: chatbot
world_module: parlai.chat_service.tasks.chatbot.worlds
overworld: MessengerOverworld
max_workers: 30
opt:
  debug: True
  models:
    blenderbot2_400M:
      model:  projects.blenderbot2.agents.blenderbot2:BlenderBot2FidAgent
      model_file: zoo:blenderbot2/blenderbot2_400M/model
      interactive_mode: True
      no_cuda: False
      override:
        share-search-and-memory-query-encoder: True
        query-generator-truncate: 256
        text-truncate: 256 
        label-truncate: 128
        search_server: 0.0.0.0:8080
        truncate: 256 

additional_args:
  page_id: 1 # Configure Your Own Page
```

However, 
The flags,
```yaml
  share-search-and-memory-query-encoder: True
  query-generator-truncate: 256
  text-truncate: 256 
  label-truncate: 128
```
do not work. See the logs:

```
10:50:59 | Overriding opt["search_server"] to 0.0.0.0:8080 (previously: None)
10:50:59 | Overriding opt["truncate"] to 256 (previously: 512)
10:50:59 | Using CUDA
10:50:59 | loading dictionary from /pantry/ran/nlp/benten/ParlAI/data/models/blenderbot2/blenderbot2_400M/model.dict
10:50:59 | num words = 50264
10:50:59 | BlenderBot2Fid: full interactive mode on.
10:51:09 | Creating the search engine retriever.
10:51:09 | No portocol provided, using "http://"
```


The logs should contain more overriding like what I got from interactive mode:
```bash
parlai interactive -mf zoo:blenderbot2/blenderbot2_400M/model --search_server 0.0.0.0:8080 --share-search-and-memory-query-encoder True --query-generator-truncate 256 --truncate 256 --text-truncate 256 --label-truncate 256 --checkpoint-activations true
```

```
10:51:49 | Overriding opt["model_file"] to /pantry/ran/nlp/benten/ParlAI/data/models/blenderbot2/blenderbot2_400M/model (previously: /checkpoint/kshuster/projects/knowledge_bot/kbot_memfix_sweep25_Fri_Jul__9/338/model.oss)
10:51:49 | Overriding opt["search_server"] to 0.0.0.0:8080 (previously: None)
10:51:49 | Overriding opt["share_search_and_memory_query_encoder"] to True (previously: False)
10:51:49 | Overriding opt["query_generator_truncate"] to 256 (previously: -1)
10:51:49 | Overriding opt["truncate"] to 256 (previously: 512)
10:51:49 | Overriding opt["text_truncate"] to 256 (previously: 512)
10:51:49 | Overriding opt["label_truncate"] to 256 (previously: 128)
10:51:49 | Using CUDA
```

**Comment by klshuster** [comment]
perhaps try using `_` instead of `-` in the config file for those parameters?

**Comment by WindChimeRan** [comment]
It works! Thanks!
```yaml
tasks:
  default:
    onboard_world: MessengerBotChatOnboardWorld
    task_world: MessengerBotChatTaskWorld
    timeout: 18000
    agents_required: 1
task_name: chatbot
world_module: parlai.chat_service.tasks.chatbot.worlds
overworld: MessengerOverworld
max_workers: 30
opt:
  debug: True
  models:
    blenderbot2_400M:
      model:  projects.blenderbot2.agents.blenderbot2:BlenderBot2FidAgent
      model_file: zoo:blenderbot2/blenderbot2_400M/model
      interactive_mode: True
      no_cuda: False
      override:
        share_search_and_memory_query_encoder: True
        query_generator_truncate: 256
        label_truncate: 128
        text_truncate: 256
        search_server: 0.0.0.0:8080
        truncate: 256
additional_args:
  page_id: 1 # Configure Your Own Page
```

**Comment by sjscotti** [comment]
Hi
I have a similar problem with GPU memory, and I tried the use  `--checkpoint-activations true` flag to lower the memory footprint during training.  However, the flag wasn't recognized.  I did some hunting around and found that it was added to a later version of parlai than what I am running.  Unless there is a simpler way to get this capability, I assume that I have to upgrade to the latest version.  But I have seen no instructions on how to do that.  Do I need to recreate my installation from scratch?  Is there an easier way to upgrade to the latest version of parlai?
Thanks!
